### DISCUSSION TITLE
Steady-state offset in tracking MPC under the influence of disturbances

### DISCUSSION THREAD

**Discussion by cezarrosa** [do-mpc/do-mpc#232] [Q&A] [answered]
I am running a simple linear problem using robust MPC. I set up 3 uncertainty values and 3 scenarios for each. If I run the simulator such as the CSTR example where the uncertainties are changed at every time step using a random generator it works just fine. However if I run the simulator with a fixed scenario the controller is unable to identify the scenario and, therefore, there is always a offset between the setpoint and the simulated value, and the controller takes no action.
I am not sure if I am doing something wrong or if it is a controller limitation or a bug.
Here are the results:
![Figure_1](https://user-images.githubusercontent.com/89585842/151160081-dd75ca9d-b0a3-4cee-8571-444129a13ded.png)

I would really appreciate your help.
Thanks in advance

**Comment by 4flixt** [comment]
Dear @cezarrosa,

thank you for your question. This doesn't look like a bug to me but I think you have slight misconception of what the robust multi-stage controller is actually doing.
The controller predicts possible future scenarios but it has no way of knowing, which of them is truly going to happen. It does not estimate the true (current) value of the uncertain parameter. To do that you must extend your MPC formulation (see for example 1.5 in this [book](https://sites.engineering.ucsb.edu/~jbraw/mpc/MPC-book-2nd-edition-3rd-printing.pdf)) or implement a parameter estimation.

That being said, I see another potential issue in your case. Notice that multi-stage MPC must choose **the same input** for all scenarios at the first time-step and will choose it such that all scenarios are feasible. Only at the next time-step (one step ahead in the future) you can counteract the effect of the different realization of the uncertainties by choosing different inputs. You can see this clearly in your plot. So the problem might be that your different scenarios for "Chloreto" spread a lot. Do you have a constraint for this state at ~3 by any chance? Because one of the scenarios is getting quite close to this value. Consequently, the nominal MPC solution must stay far away from such a hypothetical constraint - potentially making it impossible to reach your desired set-point.

Hope this helps already. Oh and please post questions in the [discussion](https://github.com/do-mpc/do-mpc/discussions) tab the next time. Issues are for bugs / feature requests etc. only.

Best,
4flixt

**Reply by cezarrosa** [reply]
Dear @4flixt,
Thanks for your reply. My apologies for posting the question in the wrong section. 
I have not imposed any constraints for "Cloreto" only for the manipulated variable "Vazão hipo" which has a upper boundary of 200.
I was expecting the controller to correct the manipulated variable based on the simulator feedback. Although it shows multiple scenarios in the plot, it seems that it is using a fixed scenario from the beginning to end of the simulation, even tough the feedback is different from the expected one.
I will attach the code here. I would be thankful if you could take a look at it. 
Thanks a lot
[MPC.zip](https://github.com/do-mpc/do-mpc/files/7971384/MPC.zip)

**Comment by 4flixt** [comment, marked as answer]
Dear @cezarrosa,

I am not sure what you mean by "it is using a fixed scenario from beginning to end". Robust multi-stage MPC considers all future scenarios equally.  It does not "know" which one is the true scenario. The input is thus chosen as a compromise between all possible future scenarios. Probably some scenarios expect a major increase in the state you want to track while other expect a minor decrease. 
The "compromise" is the trajectory you see. 

As I said before, you might need to look up some control fundamentals (like in the book that I have recommended to you ...). The problem of tracking offsets if your system is subject to disturbances is well known ... 

In any case, here is a quick solution that works for your problem:

1. Introduce a new error state in the model
```python
err = model.set_variable(var_type='_x', var_name='err', shape=(1,1))
model.set_rhs('err', Cl_ppm-Cl_ppm_sp)
```
2. Append the objective to minimize the error to the cost function:
```python
mterm = (_x['Cl_ppm'] - _tvp['Cl_ppm_sp'])**2 + _x['err']**2
lterm = (_x['Cl_ppm'] - _tvp['Cl_ppm_sp'])**2 + _x['err']**2
```
Roughly speaking, this works like an additional integral control action and it produces the following result:
<img width="1189" alt="Screenshot 2022-01-31 at 18 57 51" src="https://user-images.githubusercontent.com/32864011/151847410-cf886c3f-7206-4c7a-bb54-f004791fd3a1.png">

**Reply by cezarrosa** [reply]
Dear @4flixt,
Thanks a lot for the reply. That really helps. I will definitely read the book you sent me. Thanks for that as well.
I guess I had a misconception of how the robust MPC controller works.   
Thanks again,
Best

**Comment by 4flixt** [comment]
Btw, I have renamed your question such that others can find my answer more easily.
